montblanc/impl/rime/tensorflow/rime_ops/zernike_op_cpu.cpp

```diff
@@ -0,0 +1,150 @@
+        "noll_index must have shape [None, None, None, 4] but is " +
+        c->DebugString(in_noll_index));
+    // Assert 'noll_index' dimension '3' size
+    TF_RETURN_WITH_CONTEXT_IF_ERROR(c->WithValue(c->Dim(in_noll_index, 3), 4, &d),
+        "noll_index must have shape [None, None, None, 4] but is " +
+        c->DebugString(in_noll_index));
+    
+    // TODO. Check shape and dimension sizes for 'pointing_error'
+    ShapeHandle in_pointing_error = c->input(3);
+    // Assert 'pointing_error' number of dimensions
+    TF_RETURN_WITH_CONTEXT_IF_ERROR(c->WithRank(in_pointing_error, 5, &input),
+        "pointing_error must have shape [None, None, None, 4, 2] but is " +
+        c->DebugString(in_pointing_error));
+    // Assert 'pointing_error' dimension '3' size
+    TF_RETURN_WITH_CONTEXT_IF_ERROR(c->WithValue(c->Dim(in_pointing_error, 3), 4, &d),
+        "pointing_error must have shape [None, None, None, 4, 2] but is " +
+        c->DebugString(in_pointing_error));
+    // Assert 'pointing_error' dimension '4' size
+    TF_RETURN_WITH_CONTEXT_IF_ERROR(c->WithValue(c->Dim(in_pointing_error, 4), 2, &d),
+        "pointing_error must have shape [None, None, None, 4, 2] but is " +
+        c->DebugString(in_pointing_error));
+    
+    // TODO. Check shape and dimension sizes for 'antenna_scaling'
+    ShapeHandle in_antenna_scaling = c->input(4);
+    // Assert 'antenna_scaling' number of dimensions
+    TF_RETURN_WITH_CONTEXT_IF_ERROR(c->WithRank(in_antenna_scaling, 4, &input),
+        "antenna_scaling must have shape [None, None, 4, 2] but is " +
+        c->DebugString(in_antenna_scaling));
+    // Assert 'antenna_scaling' dimension '2' size
+    TF_RETURN_WITH_CONTEXT_IF_ERROR(c->WithValue(c->Dim(in_antenna_scaling, 2), 4, &d),
+        "antenna_scaling must have shape [None, None, 4, 2] but is " +
+        c->DebugString(in_antenna_scaling));
+    // Assert 'antenna_scaling' dimension '3' size
+    TF_RETURN_WITH_CONTEXT_IF_ERROR(c->WithValue(c->Dim(in_antenna_scaling, 3), 2, &d),
+        "antenna_scaling must have shape [None, None, 4, 2] but is " +
+        c->DebugString(in_antenna_scaling));
+    
+    
+
+    // TODO: Supply a proper shapes for output variables here,
+    // usually derived from input shapes
+    // ShapeHandle output_1 = c->MakeShape({
+    //      c->Dim(input_1, 0),  // input_1 dimension 0
+    //      c->Dim(input_2, 1)}); // input_2 dimension 1""")
+
+    ShapeHandle out_zernike_value = c->MakeShape({ 
+        c->Dim(in_coords, 0), 
+        c->Dim(in_pointing_error, 0), 
+        c->Dim(in_coeffs, 0), 
+        c->Dim(in_coeffs, 1), 
+        4 });
+    
+    c->set_output(0, out_zernike_value);
+    
+
+    // printf("output shape %s\\n", c->DebugString(out).c_str());;
+
+    return Status::OK();
+};
+
+// Register the Zernike operator.
+REGISTER_OP("Zernike")
+    .Input("coords: FT")
+    .Input("coeffs: CT")
+    .Input("noll_index: int32")
```

Comment:
The following input is also required:

```
    .Input("parallactic_angles: FT")
```

It introduces a rotation of the lm coordinate within the beam. You can achieve this easily by reproducing the code in the ebeam operator.

Reply:
Yes, good catch

1. Correct, schema is `(ntime, na)`
2. The sincos decomposition is handled by a separate operator.

Reply:
Is there a way then to call the sincos decomposition operator from within the operator? Or do I call it from outside the operator and pass it into this operator as a separate parallactic_angle_sin and a parallactic_angle_cos?